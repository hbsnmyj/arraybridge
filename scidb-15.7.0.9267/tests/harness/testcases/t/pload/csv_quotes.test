# Test the quoting of quotes in CSV files
--setup

create array simple1 <s1:string null,n:int64 null>[i=0:*,4,0]
create array simple2 <s1:string null,n:int64 null>[i=0:*,4,0]

--test

# TSV faithfully records the quotes in a standardized manner, so load
# the CSV, dump it as TSV, and see what you get....

--shell --command "printf "\"Double 'quotes' w/ , & (\"\") inside\",1\n" > /tmp/quotes1.csv"
load(simple1, '/tmp/quotes1.csv', -2, 'csv:d')
--start-igdata
save(simple1, '/tmp/quotes1.tsv', -2, 'tsv')
--stop-igdata
--shell --store --command "sed 1d /tmp/quotes1.tsv | od -c"

--echo Same but w/o the :d option, auto-quoting should handle it OK.
--shell --command "printf "\"Double 'quotes' w/ , & (\"\") inside\",1\n" > /tmp/quotes1.csv"
load(simple1, '/tmp/quotes1.csv', -2, 'csv')
--start-igdata
save(simple1, '/tmp/quotes1.tsv', -2, 'tsv')
--stop-igdata
--shell --store --command "sed 1d /tmp/quotes1.tsv | od -c"

--shell --command "printf "\"Double 'quotes' w/ , & (\\\") inside\",1\n" > /tmp/quotes2.csv"
load(simple1, '/tmp/quotes2.csv', -2, 'csv:d')
--start-igdata
save(simple1, '/tmp/quotes2.tsv', -2, 'tsv')
--stop-igdata
--shell --store --command "sed 1d /tmp/quotes2.tsv | od -c"

--echo Same but w/o the :d option, auto-quoting should handle it OK.
--shell --command "printf "\"Double 'quotes' w/ , & (\\\") inside\",1\n" > /tmp/quotes2.csv"
load(simple1, '/tmp/quotes2.csv', -2, 'csv')
--start-igdata
save(simple1, '/tmp/quotes2.tsv', -2, 'tsv')
--stop-igdata
--shell --store --command "sed 1d /tmp/quotes2.tsv | od -c"

--shell --command "printf "'Single \"quotes\" w/ , & ('') inside',3\n" > /tmp/quotes3.csv"
load(simple2, '/tmp/quotes3.csv', -2, 'csv')
--start-igdata
save(simple2, '/tmp/quotes3.tsv', -2, 'tsv')
--stop-igdata
--shell --store --command "sed 1d /tmp/quotes3.tsv | od -c"

--shell --command "printf "'Single \"quotes\" w/ , & (\\') inside',4\n" > /tmp/quotes4.csv"
load(simple2, '/tmp/quotes4.csv', -2, 'csv')
--start-igdata
save(simple2, '/tmp/quotes4.tsv', -2, 'tsv')
--stop-igdata
--shell --store --command "sed 1d /tmp/quotes4.tsv | od -c"

--echo This input used to crash in the CsvParser, but now it should just make a shadow entry.
--shell --command "printf "'Single \"quotes\" w/ , MISSING_END_QUOTE,5\n" > /tmp/quotes5.csv"
load(simple2, '/tmp/quotes5.csv', -2, 'csv', 100, simp2_shadow)
--start-igdata
save(simple2, '/tmp/quotes5.tsv', -2, 'tsv')
--stop-igdata
--shell --store --command "sed 1d /tmp/quotes5.tsv | od -c"
scan(simp2_shadow)

# Ticket 4125.
--echo This used to cause a CsvParser assertion failure.
--shell --command "printf "'Single \"quotes\" w/ , MISSING_END_QUOTE,5\n" > /tmp/quotes6.csv"
# ...and append about 8KiB more... well beyond the parser's read buffer size.
--shell --command "for x in $(seq 420) ;do echo '20 bytes of fun.,42' ;done >> /tmp/quotes6.csv"
--start-igdata
load(simple2, '/tmp/quotes6.csv', -2, 'csv', 100, simp3_shadow)
--stop-igdata
scan(simp3_shadow)

--cleanup
--shell --command "rm -f /tmp/quotes[0-9].?sv"
remove(simple1)
remove(simple2)
remove(simp2_shadow)
remove(simp3_shadow)
