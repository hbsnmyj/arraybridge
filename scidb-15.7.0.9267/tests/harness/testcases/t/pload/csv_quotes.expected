Query was executed successfully

Query was executed successfully

{i} s1,n
{0} 'Double \'quotes\' w/ , & (") inside',1

[Query was executed successfully, ignoring data output by this query.]

0000000   D   o   u   b   l   e       '   q   u   o   t   e   s   '    
0000020   w   /       ,       &       (   "   )       i   n   s   i   d
0000040   e  \t   1  \n
0000044

Same but w/o the :d option, auto-quoting should handle it OK.
{i} s1,n
{0} 'Double \'quotes\' w/ , & (") inside',1

[Query was executed successfully, ignoring data output by this query.]

0000000   D   o   u   b   l   e       '   q   u   o   t   e   s   '    
0000020   w   /       ,       &       (   "   )       i   n   s   i   d
0000040   e  \t   1  \n
0000044

{i} s1,n
{0} 'Double \'quotes\' w/ , & (") inside',1

[Query was executed successfully, ignoring data output by this query.]

0000000   D   o   u   b   l   e       '   q   u   o   t   e   s   '    
0000020   w   /       ,       &       (   "   )       i   n   s   i   d
0000040   e  \t   1  \n
0000044

Same but w/o the :d option, auto-quoting should handle it OK.
{i} s1,n
{0} 'Double \'quotes\' w/ , & (") inside',1

[Query was executed successfully, ignoring data output by this query.]

0000000   D   o   u   b   l   e       '   q   u   o   t   e   s   '    
0000020   w   /       ,       &       (   "   )       i   n   s   i   d
0000040   e  \t   1  \n
0000044

{i} s1,n
{0} 'Single "quotes" w/ , & (\') inside',3

[Query was executed successfully, ignoring data output by this query.]

0000000   S   i   n   g   l   e       "   q   u   o   t   e   s   "    
0000020   w   /       ,       &       (   '   )       i   n   s   i   d
0000040   e  \t   3  \n
0000044

{i} s1,n
{0} 'Single "quotes" w/ , & (\') inside',4

[Query was executed successfully, ignoring data output by this query.]

0000000   S   i   n   g   l   e       "   q   u   o   t   e   s   "    
0000020   w   /       ,       &       (   '   )       i   n   s   i   d
0000040   e  \t   4  \n
0000044

This input used to crash in the CsvParser, but now it should just make a shadow entry.
{i} s1,n
{0} 'Single "quotes" w/ , MISSING_END_QUOTE,5
',null

[Query was executed successfully, ignoring data output by this query.]

0000000   S   i   n   g   l   e       "   q   u   o   t   e   s   "    
0000020   w   /       ,       M   I   S   S   I   N   G   _   E   N   D
0000040   _   Q   U   O   T   E   ,   5   \   n  \t   \   N  \n
0000056

{i} s1,n,row_offset
{0} null,'Too few input record fields at position 42, line 1, column 1',42

This used to cause a CsvParser assertion failure.
[Query was executed successfully, ignoring data output by this query.]

{i} s1,n,row_offset
{0} 'Unbalanced quote in CSV input at or near record 0','Too few input record fields at position 42, line 1, column 1',0

Query was executed successfully

Query was executed successfully

Query was executed successfully

Query was executed successfully

